### PR TITLE
refactor log scale to use numeric transforms

### PR DESCRIPTION
**Pull request by EmmadUsmani**
Closes
https://github.com/metabase/metabase/issues/38780
https://github.com/metabase/metabase/issues/40212

### Description

Refactors the log y-axis scale to use our own transformation, like we do for the power scale, in order to fix the aforementioned issues.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Comment by EmmadUsmani** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/metabase/metabase/40599?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#40599** <a href="https://app.graphite.dev/github/pr/metabase/metabase/40599?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#40413** <a href="https://app.graphite.dev/github/pr/metabase/metabase/40413?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#40331** <a href="https://app.graphite.dev/github/pr/metabase/metabase/40331?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `echarts`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @EmmadUsmani and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: frontend/src/metabase/visualizations/echarts/cartesian/model/dataset.ts [hunk 1]
```diff
@@ -373,6 +423,66 @@ function filterNullDimensionValues(dataset: ChartDataset) {
...
+  dataset.forEach(datum => {
+    const datumNumericValues = seriesDataKeys
+      .map(key => getNumberOr(datum[key], null))
+      .filter(isNotNull);
+
+    const hasPositive = datumNumericValues.some(value => value > 0);
+    const hasNegative = datumNumericValues.some(value => value < 0);
+
+    if (hasPositive && hasNegative) {
+      throw Y_AXIS_CROSSING_ERROR;
```
Comment (alxnddr) [inline frontend/src/metabase/visualizations/echarts/cartesian/model/dataset.ts:447]: The old code threw strings, let's change it to throwing errors. Afair the code that handles it supports showing error messages from `error.message`

**Review by alxnddr — APPROVED** [review]
Looks good! Please have a lok at one minor comment before merging

File: frontend/src/metabase/visualizations/echarts/cartesian/model/dataset.ts (hunk 1 above)
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/echarts/cartesian/model/dataset.ts:447]: Aren't we already throwing an error here? Y_AXIS_CROSSING_ERROR is an error object, not a string

Comment (alxnddr) [reply frontend/src/metabase/visualizations/echarts/cartesian/model/dataset.ts:447]: My brain glitched, never mind 👍

**Pull request merged**